enigma/src/test/java/org/quiltmc/enigma/translation/mapping/EntryRemapperTest.java

```diff
@@ -0,0 +1,122 @@
+		remapper = project.getRemapper();
+	}
+
+	@BeforeEach
+	public void beforeEach() {
+		EntryTree<EntryMapping> mappings = new HashEntryTree<>();
+		project.setMappings(mappings, ProgressListener.none());
+		remapper = project.getRemapper();
+	}
+
+	private static void assertName(Entry<?> entry, String expected) {
+		Entry<?> deobf = remapper.deobfuscate(entry);
+		Assertions.assertNotNull(deobf);
+		Assertions.assertEquals(expected, deobf.getName());
+	}
+
+	@Test
+	public void testUnionRename() {
+		var name = "unionAB";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("e", "a", "()V"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("e", "a", "()V"), name);
+		assertName(TestEntryFactory.newMethod("a", "a", "()V"), name);
+		assertName(TestEntryFactory.newMethod("b", "a", "()V"), name);
+
+		name = "unionBC";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("f", "a", "()Z"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("f", "a", "()Z"), name);
+		assertName(TestEntryFactory.newMethod("b", "a", "()Z"), name);
+		assertName(TestEntryFactory.newMethod("c", "a", "()Z"), name);
+
+		name = "unionA3";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("g", "a", "()D"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("g", "a", "()D"), name);
+		assertName(TestEntryFactory.newMethod("a", "a", "()D"), name);
+	}
+
+	@Test
+	public void testElementRename() {
+		var name = "unionAB";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("e", "a", "()V"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("a", "a", "()V"), name);
+		assertName(TestEntryFactory.newMethod("e", "a", "()V"), name);
+		assertName(TestEntryFactory.newMethod("b", "a", "()V"), name);
+
+		name = "unionBC";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("b", "a", "()Z"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("b", "a", "()Z"), name);
+		assertName(TestEntryFactory.newMethod("f", "a", "()Z"), name);
+		assertName(TestEntryFactory.newMethod("c", "a", "()Z"), name);
+
+		name = "unionA3";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("a", "a", "()D"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("a", "a", "()D"), name);
+		assertName(TestEntryFactory.newMethod("g", "a", "()D"), name);
+
+		name = "unionCD";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("c", "a", "()F"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("c", "a", "()F"), name);
+		assertName(TestEntryFactory.newMethod("d", "a", "()F"), name);
+
+		name = "unionDC";
+		remapper.putMapping(newVC(), TestEntryFactory.newMethod("d", "a", "()F"), new EntryMapping(name));
+
+		assertName(TestEntryFactory.newMethod("d", "a", "()F"), name);
+		assertName(TestEntryFactory.newMethod("c", "a", "()F"), name);
+	}
+
+	private static ValidationContext newVC() {
```

Comment:
this stuff should probably be moved to our common testing utils since it's in multiple files now (also in `TestMappingValidator`)